lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetColumnChunkIterator.java

```diff
@@ -13,112 +13,104 @@
  */
 package io.trino.parquet.reader;
 
-import io.airlift.slice.BasicSliceInput;
-import io.airlift.slice.Slice;
 import io.trino.parquet.DataPage;
 import io.trino.parquet.DataPageV1;
 import io.trino.parquet.DataPageV2;
 import io.trino.parquet.DictionaryPage;
 import io.trino.parquet.ParquetCorruptionException;
 import org.apache.parquet.column.Encoding;
-import org.apache.parquet.column.statistics.Statistics;
 import org.apache.parquet.format.DataPageHeader;
 import org.apache.parquet.format.DataPageHeaderV2;
 import org.apache.parquet.format.DictionaryPageHeader;
 import org.apache.parquet.format.PageHeader;
 import org.apache.parquet.format.Util;
 import org.apache.parquet.internal.column.columnindex.OffsetIndex;
 
+import javax.annotation.Nullable;
+
 import java.io.IOException;
-import java.util.LinkedList;
-import java.util.List;
+import java.util.Iterator;
 import java.util.Optional;
 import java.util.OptionalLong;
 
-import static com.google.common.base.Verify.verify;
+import static com.google.common.base.Preconditions.checkArgument;
 import static io.trino.parquet.ParquetTypeUtils.getParquetEncoding;
 import static java.util.Objects.requireNonNull;
+import static org.apache.parquet.format.PageType.DICTIONARY_PAGE;
 
-public class ParquetColumnChunk
+public final class ParquetColumnChunkIterator
+        implements Iterator<DataPage>
 {
     private final Optional<String> fileCreatedBy;
     private final ColumnChunkDescriptor descriptor;
-    private final List<Slice> slices;
-    private int sliceIndex;
-    private BasicSliceInput input;
+    private ChunkedInputStream input;
     private final OffsetIndex offsetIndex;
 
-    public ParquetColumnChunk(
+    @Nullable
+    private DictionaryPage dictionaryPage;
```

Comment:
I think we should try to avoid holding reference to `DictionaryPage` here and in PageReader. While this reference exists, the JVM can't release the entire underlying first 8MB chunk that this is read from.
We can make this class an iterator on `Page` and let DictionaryPage be the first page in that. This class can have a `hasDictionary` method instead of `getDictionaryPage`.
`PageReader#readDictionary` can pull first page from iterator if we have dictionary, return null otherwise. We can use some checkStates in PageReader to ensure that `readDictionary` is always called first.

Reply:
`PageReader` holds a reference to the `DictionaryPage` before this pr.
How about doing this as a follow-up to limit the scope here?

Reply:
We should at least avoid holding DictionaryPage reference in this class as that was not done previously in ParquetColumnChunk. If you do that then the remaining change in PageReader is small.

Reply:
I did the full change but this looks like we should merge PageReader and ParquetColumnChunkIterator.

Reply:
I'm thinking we can keep them separate until it becomes really necessary to merge, because it's easier to feed pages to PageReader in tests rather than having to provide chunks.

Reply:
>  it's easier to feed pages to PageReader in tests rather than having to provide chunks

Should be easy to create chunks based on the same data that goes to the DataPage. e.g. add page header and concatenate the slices.